core/go/internal/sequencer/coordinator/select_transaction.go

```diff
@@ -0,0 +1,206 @@
+	log.L(ctx).Trace("selecting next transaction to assemble")
+	txn, err := c.transactionSelector.SelectNextTransactionToAssemble(ctx, event)
+	if err != nil {
+		log.L(ctx).Errorf("error selecting transaction: %v", err)
+		return err
+	}
+	if txn == nil {
+		log.L(ctx).Info("no transaction found to process")
+		return nil
+	}
+
+	transactionSelectedEvent := &transaction.SelectedEvent{}
+	transactionSelectedEvent.TransactionID = txn.ID
+	err = txn.HandleEvent(ctx, transactionSelectedEvent)
+	return err
+
+}
+
+func (c *coordinator) AddTransactionToBackOfPool(ctx context.Context, txn *transaction.Transaction) {
+	c.pooledTransactionsMutex.Lock()
+	defer c.pooledTransactionsMutex.Unlock()
+	c.pooledTransactions = append(c.pooledTransactions, txn)
+}
+
+/* Functions of the TransactionPool interface required by the transactionSelector */
+func (c *coordinator) GetNextPooledTransaction(ctx context.Context) *transaction.Transaction {
+	c.pooledTransactionsMutex.Lock()
+	defer c.pooledTransactionsMutex.Unlock()
+	if len(c.pooledTransactions) == 0 {
+		return nil
+	}
+	nextPooledTx := c.pooledTransactions[0]
+	c.pooledTransactions = c.pooledTransactions[1:]
+	return nextPooledTx
+}
+
+func (c *coordinator) GetTransactionByID(_ context.Context, txnID uuid.UUID) *transaction.Transaction {
+	return c.transactionsByID[txnID]
+}
+
+// define the interface for the transaction selector algorithm
+// we inject into the algorithm a dependency that allows it to query the state of the transaction pool under the control of the coordinator
+type TransactionSelector interface {
+	SelectNextTransactionToAssemble(ctx context.Context, event *TransactionStateTransitionEvent) (*transaction.Transaction, error)
+}
+
+// define the interface that the transaction selector algorithm uses to query the state of the transaction pool
+type TransactionPool interface {
+	// Put this transaction to the back of the pooled queue
+	AddTransactionToBackOfPool(ctx context.Context, txn *transaction.Transaction)
+	// Return the next transaction that is pooled, in FIFO order (i.e. from the front of the queue/slice)
+	GetNextPooledTransaction(ctx context.Context) *transaction.Transaction
+
+	GetTransactionByID(ctx context.Context, txnID uuid.UUID) *transaction.Transaction
+}
+
+type transactionSelector struct {
+	transactionPool             TransactionPool
+	currentAssemblingOriginator *originatorLocator
+}
+
+func NewTransactionSelector(ctx context.Context, transactionPool TransactionPool) TransactionSelector {
+
+	selector := &transactionSelector{
+		transactionPool:             transactionPool,
+		currentAssemblingOriginator: nil,
+	}
+
+	return selector
+}
+
+type originatorLocator struct {
+	node     string
+	identity string
+	txID     uuid.UUID
+}
+
+// Determine if there is still an active assembler, so we know if it's safe to select a new transaction to start processing
```

Comment:
Have really struggled to read this function.

Title suggest it's asking a question `transactionStillBeingAssembled`?.

Function has a `TODO` to change it - which sound very true... because it's function seems to be able trying to set a variable `currentAssemblingOriginator` to `nil` based on processing a `TransactionStateTransitionEvent`.

So overall it feels like this function is a placeholder for a core part of the state machine, which is processing state transition event.